sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientRegistration.cs

```diff
@@ -63,5 +70,65 @@ public TClient GetClient(IServiceProvider serviceProvider, object options, Token
                 return _cachedClient;
             }
         }
+
+        public async ValueTask DisposeAsync()
+        {
+            if (_clientInitialized)
+            {
+                if (_asyncDisposable)
+                {
+                    IAsyncDisposable disposableClient;
+
+                    lock (_cacheLock)
+                    {
+                       if (!_clientInitialized)
+                       {
+                           return;
+                       }
+
+                       disposableClient = (IAsyncDisposable)_cachedClient;
+
+                       _cachedClient = default;
+                       _clientInitialized = false;
+                    }
+
+                    await disposableClient.DisposeAsync().ConfigureAwait(false);
+                }
+                else if (_disposable)
+                {
+                    Dispose();
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_clientInitialized)
+            {
+                if (_disposable)
+                {
+                    IDisposable disposableClient;
+
+                    lock (_cacheLock)
+                    {
+                       if (!_clientInitialized)
+                       {
+                           return;
+                       }
+
+                       disposableClient = (IDisposable)_cachedClient;
+
+                       _cachedClient = default;
+                       _clientInitialized = false;
+                    }
+
+                    disposableClient.Dispose();
+                }
+                else if (_asyncDisposable)
+                {
+                    DisposeAsync().GetAwaiter().GetResult();
```

Comment:
I'm not sure if it is correct to call DisposeAsync from Dispose - https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#implement-both-dispose-and-async-dispose-patterns

Reply:
The case that I'm shooting for here is "we only implement DisposeAsync and not Dispose, but we still want to be cleaned up."    This isn't a problem in newer versions of the .NET DI library, where the service factory implements both, but in earlier versions - like the one we reference - they only support Dispose.    This would leave clients like Event Hubs or Service Bus unable to be cleaned up since we don't implement Dispose.

Reply:
Discussed offline; decided to keep the implementation.